src/Logging/test/ConsoleLoggerTest.cs

```diff
@@ -377,19 +378,65 @@ public void WriteCore_LogsCorrectTimestamp(ConsoleLoggerFormat format, LogLevel
             switch (format)
             {
                 case ConsoleLoggerFormat.Default:
-                {
-                    Assert.Equal(3, sink.Writes.Count);
-                    Assert.Matches("^\\d{14}\\s$", sink.Writes[0].Message);
-                    Assert.StartsWith(levelPrefix, sink.Writes[1].Message);
-                }
-                break;
+                    {
+                        Assert.Equal(3, sink.Writes.Count);
+                        Assert.StartsWith(levelPrefix, sink.Writes[1].Message);
+                        Assert.Matches("^\\d{4}\\D\\d{2}\\D\\d{2}\\D\\d{2}\\D\\d{2}\\D\\d{2}\\D\\d{2}\\s$", sink.Writes[0].Message);
+                        var parsedDateTime = DateTimeOffset.Parse(sink.Writes[0].Message.Trim());
+                        Assert.NotEqual(0, parsedDateTime.Offset.Hours);
```

Comment:
This causes a [test-failure](https://dev.azure.com/dnceng/public/_build/results?buildId=370092&view=ms.vss-test-web.build-test-results-tab&runId=11249960&resultId=100145&paneView=debug).

Reply:
Yes, see my comment here: https://github.com/aspnet/Extensions/pull/1896#discussion_r329352016

It works locally but my guess is that the build agent runs in UTC local which makes it fail. Ideally the tests sets the locale to a non-UTC one to enforce this but seeking help for this as I'm not an expert on that topic.

Reply:
Oh, you're fast, I'm still thinking for a good reliable approach 😃

Reply:
Hm, I think it either boils down to https://github.com/aspnet/Extensions/pull/1896#discussion_r300340699 or to add an `IDateTimeProvider` that is injected to the console-logger and can be mocked.
To change the current timezone can't be done with (pure) .NET, and I think we shouldn't change the system's timezone just for this test either.

The easiest would be to add `IDateTimeProvider`, but this comes with the cost of another level of indirection (and an interface-call -- which may harm perf, I have an issue with perf of console-logger in mind).
Perf-wise instead of the interface and abstract class (with sealed overrides in the concrete implementation) could be used.
Just add an ctor-overload (internal) which accepts an instance of `IDateTimeProvider`, that can be used for testing.
The default ctor uses a concrete default implementation.

That's quite some change in the code-base. If there are other places where this could be used to, I think it's worth it (maybe there's already such an abstraction lying around, didn't check that).

Otherwise just go with the first mentioned approach. To address your concern "who says that method is being called?" you could
* just rely on review :wink:
* for debug build implement a callback which you check, so like
```c#
// ConsoleLogger
private static string GetTimestamp(string timestampFormat, bool useUtcTimezone)
    => GetTimestamp(timestampFormat, useUtcTimezone, DateTime.Now);

#if DEBUG
public static Action GetTimestampCalled;
#endif

internal static string GetTimestamp(string timestampFormat, bool useUtcTimezone, DateTime now)
{
#if DEBUG
    GetTimestampCalled?.Invoke();
#endif

    if (timestampFormat == null)
        return null;

    if (useUtcTimezone)
        now = now.ToUniversalTime();

    return now.ToString(timestampFormat);
}
```
```c#
// Test, only run in DEBUG
    bool getTimestampCalled = false;
    void OnGetTimestampCalled() => getTimestampCalled = true;

    ConsoleLogger.GetTimestampCalled += OnGetTimestampCalled;
    try
    {
        string actual = logger.Log();

        Assert.NotNull(actual);
        Assert.True(getTimestampCalled);
    }
    finally
    {
        // de-registration may not be necessary in tests, but it prevents memory leaks (static event)
        ConsoleLogger.GetTimestampCalled -= OnGetTimestampCalled;
    }
```

I'm not absolutely sure what approach is better.

Reply:
This feels a bit odd to me given we're changing the library just for sake of testing but if that's what you guys want, I can do that.

That said, I can take a crack at the `IDateTimeProvider` approach and see how that goes, if you want.

Reply:
Even with the `IDateTimeProvider`-approach outlined above, we don't have a guarantee it is called (except we use a mock).
Now I think this will be over-engineering for this "simple" test. Right now I'll just go with https://github.com/aspnet/Extensions/pull/1896#discussion_r300340699, as it is quite simple (in code and tests).

PS: I'm not on the team, so I don't know what they want either :wink:

Reply:
> It works locally but my guess is that the build agent runs in UTC local which makes it fail

All our build agents run in Azure, which sets the local time zone to UTC. We also need the tests to be reliable in places that do that, so we need them to pass on UTC time zones.

This is indeed a challenging one to automate, especially since all our CI automation runs in UTC already so toggling this option is a no-op. Changing the locale of the machine during the build is error-prone at best.

We do have an [`ISystemClock` abstraction we've used in places](https://github.com/aspnet/Extensions/blob/master/src/Caching/Abstractions/src/Internal/ISystemClock.cs). It might be worth looking at a similar approach here.

Reply:
I'll try to see if I can make some sense out of it, I'm not that much of an expert on that topic 😅 

I'm assuming you want to go the same route as above where `ISystemClock` is injected via an internal ctor or what do you have in mind exactly?

Reply:
The approach worked, thanks for the tip.

However, there is one issue with the references but I'll add a comment there.

Reply:
Actually I got confused and [`ISystemClock`](https://github.com/aspnet/Extensions/blob/master/src/Caching/Abstractions/src/Internal/ISystemClock.cs) won't cut it given it only supports `Utc` and not local time.

What I can do is extend it so that it contains a `Now` property, but wanted to check with you guys first if that's ok.

Reply:
Actually, why not just compare the offset to `DateTimeOffset.Now.Offset` (and `DateTimeOffset.UtcNow.Offset` in the other case)? Yes, it means that on our build agents it's always testing against the UTC offset but that's because our agents run with UTC as the local time zone. In other builds (local dev builds, etc.) that won't be the case.

I think that may be the best we can do without introducing a lot of extra abstractions, internal code, mocks, etc.

Reply:
Actually that's a good point - Will change this later today/tomorrow!

Reply:
Thanks! Testing things involving `DateTime` is notoriously tricky after all ;).

Reply:
My pleasure, things should have been changed now.

That is correct, but it should not be a reason for creating such a sloppy PR like this one - Sorry for that!